hw4/lebr0nli/_matrix.cpp

```diff
@@ -0,0 +1,156 @@
+    rows = other.rows;
+    cols = other.cols;
+    other.rows = other.cols = 0;
+  }
+  return *this;
+}
+
+inline double &Matrix::operator()(size_t row, size_t col) {
+  return data[row * cols + col];
+}
+
+inline const double &Matrix::operator()(size_t row, size_t col) const {
+  return data[row * cols + col];
+}
+
+bool Matrix::operator==(const Matrix &other) const {
+  if (rows != other.rows || cols != other.cols)
+    return false;
+  return data == other.data;
+}
+
+double Matrix_getitem(const Matrix &m,
+                      const std::pair<size_t, size_t> &indices) {
+  const auto [row, col] = indices;
+  if (row >= m.rows || col >= m.cols) {
+    throw std::out_of_range("Matrix indices out of range");
+  }
+  return m(row, col);
+}
+
+void Matrix_setitem(Matrix &m, const std::pair<size_t, size_t> &indices,
+                    double value) {
+  const auto [row, col] = indices;
+  if (row >= m.rows || col >= m.cols) {
+    throw std::out_of_range("Matrix indices out of range");
+  }
+  m(row, col) = value;
+}
+
+Matrix multiply_naive(const Matrix &a, const Matrix &b) {
+  if (a.cols != b.rows) {
+    throw std::invalid_argument(
+        "Matrix dimensions do not match for multiplication");
+  }
+
+  Matrix result(a.rows, b.cols);
+  for (size_t i = 0; i < a.rows; ++i) {
+    for (size_t j = 0; j < b.cols; ++j) {
+      double sum = 0.0;
+      for (size_t k = 0; k < a.cols; ++k) {
+        sum += Matrix_at(a, i, k) * Matrix_at(b, k, j);
+      }
+      Matrix_at(result, i, j) = sum;
+    }
+  }
+  return result;
+}
+
+Matrix multiply_tile(const Matrix &a, const Matrix &b, size_t tile_size) {
+  if (a.cols != b.rows) {
+    throw std::invalid_argument(
+        "Matrix dimensions do not match for multiplication");
+  }
+
+  Matrix result(a.rows, b.cols);
+
+  std::fill(result.data.begin(), result.data.end(), 0.0);
+
+  for (size_t i = 0; i < a.rows; i += tile_size) {
+    for (size_t k = 0; k < a.cols; k += tile_size) {
+      for (size_t j = 0; j < b.cols; j += tile_size) {
+        size_t i_end = std::min(a.rows, i + tile_size);
+        size_t k_end = std::min(a.cols, k + tile_size);
+        size_t j_end = std::min(b.cols, j + tile_size);
+
+        for (size_t ii = i; ii < i_end; ++ii) {
+          for (size_t kk = k; kk < k_end; ++kk) {
+            double r = Matrix_at(a, ii, kk);
+            for (size_t jj = j; jj < j_end; ++jj) {
+              Matrix_at(result, ii, jj) += r * Matrix_at(b, kk, jj);
+            }
+          }
+        }
+      }
+    }
+  }
+  return result;
+}
+
+Matrix multiply_mkl(const Matrix &a, const Matrix &b) {
+  if (a.cols != b.rows) {
+    throw std::invalid_argument(
+        "Matrix dimensions do not match for multiplication");
+  }
+
+  Matrix result(a.rows, b.cols);
+  cblas_dgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans, a.rows, b.cols, a.cols,
+              1.0, a.data.data(), a.cols, b.data.data(), b.cols, 0.0,
+              result.data.data(), result.cols);
+  return result;
+}
```

Comment:
Hi @lebr0nli ,

When using git for version control, it is recommended to add a newline at the end of files. This is because git compares files line by line, and the end of each line is identified by a newline character.

For more details, refer to the following link:
https://medium.com/@alexey.inkin/how-to-force-newline-at-end-of-files-and-why-you-should-do-it-fdf76d1d090e

Reply:
Thanks for pointing that out!
I didn't realize a single newline could make such a difference for version control. I'll make sure to add the newline and enable `InsertNewlineAtEOF` for my `clang-format` next time :)